### PR TITLE
Using va_copy to safely pass an arguments around

### PR DESCRIPTION
**Pull request by smithx**
[Justification from Julio Merino](http://julio.meroh.net/2011/09/using-vacopy-to-safely-pass-ap.html)

### CODE REVIEW

File: src/snprintf.cxx [hunk 1]
```diff
@@ -69,19 +69,29 @@ static inline
 int
 vftprintf (std::FILE * file, tchar const * fmt, std::va_list args)
 {
+    int ret;
+
+    va_list args_copy;
+
+    va_copy(args_copy, args);
```
Comment (wilx) [inline src/snprintf.cxx]: I am not convinced any of these are necessary. With your patch we already copy the `args` in caller.

File: src/snprintf.cxx [hunk 2]
```diff
@@ -95,20 +105,27 @@ vstprintf (tchar * dest, std::size_t dest_size, tchar const * fmt,
     std::va_list args)
 {
     int ret;
+
+    va_list args_copy;
+
+    va_copy(args_copy, args);
```
Comment (wilx) [inline src/snprintf.cxx]: Again, I am not convinced these are necessary.

File: src/snprintf.cxx [hunk 3]
```diff
@@ -128,35 +145,41 @@ vsntprintf (tchar * dest, std::size_t dest_size, tchar const * fmt,
 {
     int ret;
 
+    va_list args_copy;
+
+    va_copy(args_copy, args);
```
Comment (wilx) [inline src/snprintf.cxx]: Again, not convinced these are necessary.

File: src/snprintf.cxx [hunk 4]
```diff
@@ -212,23 +235,34 @@ snprintf_buf::print_va_list (tchar const * & str, tchar const * fmt,
...
+    va_end(args_copy);
+
     assert (printed == sprinted);
 
     buf[sprinted] = 0;
 
 #else
-    printed = vsntprintf (&buf[0], buf_size - 1, fmt, args);
+    va_list args_copy;
+    va_copy(args_copy, args);
```
Comment (wilx) [inline src/snprintf.cxx]: And again, I don't think these are necessary.

File: src/snprintf.cxx [hunk 5]
```diff
@@ -212,23 +235,34 @@ snprintf_buf::print_va_list (tchar const * & str, tchar const * fmt,
...
+
+    printed = vftprintf (fnull, fmt, args_copy);
     if (printed == -1)
         LogLog::getLogLog ()->error (
             LOG4CPLUS_TEXT ("Error printing into NULL_FILE."), true);
 
     buf.resize (printed + 1);
-    int sprinted = vstprintf (&buf[0], buf.size (), fmt, args);
+    int sprinted = vstprintf (&buf[0], buf.size (), fmt, args_copy);
```
Comment (wilx) [inline src/snprintf.cxx]: Reading some docs, it seems to me this should use the original `args` here.

**Comment by wilx** [comment]
Is there an actual code that does not work or crashes without these changes?

**Comment by smithx** [comment]
Yes, my project log garbage without this changes, but only on Linux

**Comment by wilx** [comment]
What platform is your Linux running on?

I am willing to accept the one `va_copy` I have already marked during review. But the rest seems unnecessary. You will have to provide more convincing arguments for the rest.

**Comment by smithx** [comment]
Debian 8 on Intel (GCC 4.9 - 7.3). I have not done any research on the location of std::va_list on the stack, only found a solution at Julio Merino's post

**Comment by wilx** [comment]
What I am getting at is that I can see the `va_list` being passed untouched through at least two calls. Somebody in vaargs function does `va_start` and the resulting `va_list` can be passed down to a function that can do something with it. E.g., in Bionic C library: https://android.googlesource.com/platform/bionic/+/ics-mr0/libc/stdio/vfprintf.c#159

**Comment by smithx** [comment]
I have no more convincing arguments.

**Comment by wilx** [comment]
> I have no more convincing arguments.

A test case would do. Also, addition or modified unit test that fails without your patch would be great. See existing unit test in [`snprintf.cxx`](https://github.com/log4cplus/log4cplus/blob/master/src/snprintf.cxx#L269).

**Comment by wilx** [comment]
Done in #438.

**Pull request closed without merging**